sig-scalability/slos/network_latency.md

```diff
@@ -0,0 +1,49 @@
+## In-cluster network latency SLIs/SLOs details
+
+### Definition
+
+| Status | SLI | SLO |
+| --- | --- | --- |
+| __WIP__ | In-cluster network latency from a single node, measured as latency of per second ping from node to "null service", measured as 99th percentile over last 5 minutes. | In default Kubernetes installataion with RTT between nodes <= Y, 99th percentile of (99th percentile over all nodes) per cluster-day <= X |
+
+### User stories
+- As a user of vanilla Kubernetes, I want some guarantee how fast my http
+request to some Kubernetes service reaches its endpoint
+
+### Other notes
+- We obviously can't give any guarantee in a general case, because cluster
+administrators may configure cluster as they want.
+- As a result, we define the SLI to be very generic (no matter how your cluster
+is set up), but we provide SLO only for default installations with an additional
+requirement that low-level RTT between nodes is lower than Y.
+- Network latency is one of the most crucial aspects from the point of view
+of application performance, especially in microservices world. As a result, to
+meet user expectations, we need to provide some guarantees arount that.
+- We decided for the SLI definition as formulated above, because:
+  - it represents a user oriented end-to-end flow - it involves among others
+    latency of in-cluster network programming mechanism (e.g. iptables). <br/>
+    __TODO:__ We considered making DNS resolution part of it, but decided not
+    to mix them. However, longer term we should consider joining them.
+  - it is easily measurable in all running clusters without any significant
+    overhead (e.g. measuring request latencies coming from all pods on a given
+    node would require some additional instrumentation, such as a side car for
+    each of them, and that overhead may be not acceptable in many cases)
+  - it is not application-specific
+
+### Caveats
+- The SLI is formulated for a single node, even though users are mostly
+interested in the aggregation across all nodes (that is done only at the SLO
+level). However, that provides very similar guarantees and makes it fairly
+easy to measure.
+- The prober reporting that is fairly trivial and itself needs only negligible
+amount of resources. However, to avoid any visible overhead in the cluster
+(in terms of additionally needed components) **we will make it part of kube-proxy
```

Comment:
This is going to test (as you stated in the SLI column) traffic from nodes.  I don't think that is the most representative or useful metric.  It seems more correct to test from a pod, through a service, to a pod, across nodes in the same primary topology.

I don't know how to generically express "same primary topology".  In GCP that is zone, but topology is arbitrary -- maybe that needs to be parameterized?

We might ALSO want to SLI pod-to-pod without a service (or with a headless service) with the same conditions.

@johnbelamaric

Reply:
> This is going to test (as you stated in the SLI column) traffic from nodes. I don't think that is the most representative or useful metric. It seems more correct to test from a pod, through a service, to a pod, across nodes in the same primary topology.

Kube-proxy is also a pod. So why this isn't what you want?

Re sam primary topology - given that we currently don't really support that fully, I'm a bit sceptical about doing that. Once we have topology-aware service routing (https://github.com/kubernetes/community/pull/2846) we can update that SLI.

@thockin - WDYT?

Reply:
Discussed that offline with @thockin 
The main points were:
- we can't use kube-proxy, because it's in host network - so I switched that to use dedicated prober
- about topologies - making that explicit that we acknowledge very different latencies if nodes are in different topologies - I added explicit point about that.